File: dp03_can_sum.py
def can_sum(target, nums):
    if not target:
        return True

    if target < 0:
        return False

    for val in nums:
        rem = target - val
        if can_sum(rem, nums):
            return True

    return False 

def can_sum_dy(target, nums, memo={}):
    if target in memo:
        return memo[target]

    if not target:
        return True

    if target < 0:
        return False

    for val in nums:
        rem = target - val
        if can_sum_dy(rem, nums, memo) == True:
            memo[target] = True
            return True

    memo[target] = False
    return False 

def main():
    # print(can_sum(7, [2, 3]))
    # print(can_sum(7, [5, 3, 4, 7]))
    # print(can_sum(7, [2, 4]))
    # print(can_sum(8, [2, 3, 5]))
    # print(can_sum(320, [7, 14]))

    print(can_sum_dy(7, [2, 3], {}))
    print(can_sum_dy(7, [5, 3, 4, 7], {}))
    print(can_sum_dy(7, [2, 4], {}))
    print(can_sum_dy(8, [2, 3, 5], {}))
    print(can_sum_dy(320, [7, 14], {}))
    
main()

File: dp03_canSum.go
package main

import "fmt"

func main() {
	// fmt.Println(canSum(7, []int{2, 3}))
	// fmt.Println(canSum(7, []int{5, 3, 4, 7}))
	// fmt.Println(canSum(7, []int{2, 4}))
	// fmt.Println(canSum(8, []int{2, 3, 5}))
	// fmt.Println(canSum(320, []int{7, 14}))

	fmt.Println(canSumDy(7, []int{2, 3}, map[int]bool{}))
	fmt.Println(canSumDy(7, []int{5, 3, 4, 7}, map[int]bool{}))
	fmt.Println(canSumDy(7, []int{2, 4}, map[int]bool{}))
	fmt.Println(canSumDy(8, []int{2, 3, 5}, map[int]bool{}))
	fmt.Println(canSumDy(320, []int{7, 14}, map[int]bool{}))
}

func canSum(targetSum int, nums []int) bool {
	if targetSum == 0 {
		return true
	}
	if targetSum < 0 {
		return false
	}

	for _, val := range nums {
		reminder := targetSum - val
		if canSum(reminder, nums) {
			return true
		}
	}

	return false
}

func canSumDy(targetSum int, nums []int, memo map[int]bool) bool {
	if val, ok := memo[targetSum]; ok {
		return val
	}

	if targetSum == 0 {
		return true
	}
	if targetSum < 0 {
		return false
	}

	for _, val := range nums {
		reminder := targetSum - val
		if canSumDy(reminder, nums, memo) {
			memo[targetSum] = true
			return true
		}
	}

	memo[targetSum] = false
	return false
}
